src/seismometer/report/fairness.py

```diff
@@ -0,0 +1,550 @@
+                return key.apply(lambda x: cohort_groups.index(x))
+            case _:
+                return key
+
+    return dataframe.sort_values(by=["Cohort", "Count"], key=fairness_sort_key)
+
+
+def fairness_table(
+    dataframe: pd.DataFrame,
+    metric_fn: Callable[..., dict[str, float]],
+    metric_list: list[str],
+    fairness_ratio: float = 0.25,
+    cohort_dict: dict[str, tuple[Any]] = None,
+    *,
+    censor_threshold: int = 10,
+    **kwargs,
+) -> HTML:
+    """
+    Fairness table for evaluating metrics across cohorts, found by taking the largest subgroup within each cohort
+    as the default cohort and taking the ratio between a subgroup's metric and the default group's metric.
+
+    For example if if a cohort has three classes A, B, and C with counts of 10, 20, and 30 respectively, the default
+    cohort would be C. For a metric M, we would calculate M(A), M(B) and M(C) and then calculate the ratios
+    M(A)/M(C) and its reciprical M(C)/M(A).
+
+    If M(A)/M(C) > 1 + limit, then cohort A will be flagged as higher than the default.
+    If M(A)/M(C) > 1 + 2 * limit, then cohort A will be flagged as critically higher than the default.
+
+    If M(C)/M(A) > 1 + limit, then cohort A will be flagged as lower than the default.
+    If M(C)/M(A) > 1 + 2 * limit, then cohort A will be flagged as critically lower than the default.
+
+    Parameters
+    ----------
+    dataframe : pd.DataFrame
+        Source data to generate a fairness table for
+    metric_fn : Callable[..., dict[str, float]]
+        Metric function to generate raw metrics, which MUST be positive values.
+    metric_list : list[str]
+        List of metrics to use from the metric function
+    fairness_ratio : float
+        Ratio of acceptable difference between cohorts, 20% is 0.2, 200% is 2.0.
+        Bound is mulitplicatively symmetric around 1, so 200% means up to 3x larger or 3x smaller (1/3 the original).
+        A typical bound is 0.25 (1.25x larger or 0.8x smaller)
+    cohort_dict : dict[str, tuple[Any]]
+        collection of cohort groups to loop over
+    censor_threshold : int, optional
+        Limit at which a cohort group will be removed from the table if not enough samples are found, by default 10.
+
+    Returns
+    -------
+    HTML
+        The HTML table for the fairness evaluation
+    """
+    fairness_groups = []
+    metric_groups = []
+    icon_groups = []
+
+    if fairness_ratio <= 0:
+        raise ValueError("Fairness ratio must be greater than 0")
+
+    if not cohort_dict:
+        raise ValueError("No cohorts provided for fairness evaluation")
+
+    for cohort_column in cohort_dict:
+        cohort_indices = []
+        cohort_values = []
+        for cohort_class in cohort_dict[cohort_column]:
+            cohort_indices.append((cohort_column, cohort_class))
+
+            cohort_filter = FilterRule.eq(cohort_column, cohort_class)
+            cohort_dataframe = cohort_filter.filter(dataframe)
+
```

Comment:
We might want to shortcircuit here if we filter down to an empty frame

Reply:
I don't want to short circuit here, I'm going to instead return default metrics when the dataframe is empty.

Reply:
Sounds good